aten/src/ATen/native/SharedReduceOps.h

```diff
@@ -80,8 +80,20 @@ struct WelfordData {
   scalar_t m2;
   index_t n;
   combine_t nf;
-  C10_HOST_DEVICE WelfordData() : mean(0), m2(0), n(0), nf(0)  {}
-  C10_DEVICE WelfordData(scalar_t mean, scalar_t m2, index_t n, combine_t nf) : mean(mean), m2(m2), n(n), nf(nf) {}
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  // Shared memory requires no initialization.
+  C10_HOST_DEVICE WelfordData() = default;
```

Comment:
Yes, but if we add dynamic initialization in ctor here, there will be a warning "dynamic initialization is not supported for a function-scope static __shared__ variable within a __device__/__global__ function".

Reply:
Sorry for delay, I'm still uneasy about this initialization, otherwise PR looks good.

Reply:
Actually there are some discussions for this before. Not sure if there is any better way to handle this.
https://github.com/pytorch/pytorch/pull/13967

Reply:
Quoting from our compiler team for a similar issue:

```
__shared__ variables cannot be initialized.
```
...
and for the case of dynamic initialization:
```
the compiler reports a warning (rather than an error, for legacy reasons). The program will build but will not execute correctly. The warning reported is:

j5.cu(6): warning: dynamic initialization is not supported for a function-scope static __shared__ variable within a __device__/__global__ function
```

Can we keep the default ctor unchanged as
```
C10_HOST_DEVICE WelfordData() : mean(0), m2(0), n(0), nf(0)  {}
```

and inside the kernel, do
```CUDA
__shared__ char val_shared_[C10_WARP_SIZE * sizeof(WelfordType)];
WelfordType *val_shared = reinterpret_cast<WelfordType *>(val_shared_);
if (threadIdx.x < 32) {
  val_shared[threadIdx.x] = WelfordType();
}
__syncthreads();
```

Reply:
Thanks for the advice.